sklearn/cluster/optics.py

```diff
@@ -0,0 +1,585 @@
+    
+    Nodelist = []
+    Nodelist.append((Node1,LocalMax1))
+    Nodelist.append((Node2,LocalMax2))
+    
+    # set a lower threshold on how small a significant maxima can be
+    significantMin = .003
+
+    if RPlot[s] < significantMin:
+        node.assignSplitPoint(-1)
+        #if splitpoint is not significant, ignore this split and continue
+        _cluster_tree(node,parentNode, localMaximaPoints, 
+                    RPlot, RPoints, min_cluster_size)
+        return
+        
+        
+    # only check a certain ratio of points in the child 
+    # nodes formed to the left and right of the maxima
+    checkRatio = .8
+    checkValue1 = int(np.round(checkRatio*len(Node1.points)))
+    checkValue2 = int(np.round(checkRatio*len(Node2.points)))
+    if checkValue2 == 0:
+        checkValue2 = 1
+    avgReachValue1 = float(np.average(RPlot[(Node1.end - 
+                                             checkValue1):Node1.end]))
+    avgReachValue2 = float(np.average(RPlot[Node2.start:(Node2.start + 
+                                                         checkValue2)]))
+
+
+    '''
+    To adjust the fineness of the clustering, adjust the following ratios.
+    The higher the ratio, the more generous the algorithm is to preserving
+    local minimums, and the more cuts the resulting tree will have.
+    '''
+
+    # the maximum ratio we allow of average height of clusters 
+    # on the right and left to the local maxima in question
+    maximaRatio = .75
+    
+    # if ratio above exceeds maximaRatio, find which of the clusters 
+    # to the left and right to reject based on rejectionRatio
+    rejectionRatio = .7
+    
+    if (float(avgReachValue1 / float(RPlot[s])) > 
+        maximaRatio or float(avgReachValue2 / 
+                             float(RPlot[s])) > maximaRatio):
+
+        if float(avgReachValue1 / float(RPlot[s])) < rejectionRatio:
+          #reject node 2
+            Nodelist.remove((Node2, LocalMax2))
+        if float(avgReachValue2 / float(RPlot[s])) < rejectionRatio:
+          #reject node 1
+            Nodelist.remove((Node1, LocalMax1))
+        if (float(avgReachValue1 / float(RPlot[s])) >= 
+            rejectionRatio and float(avgReachValue2 / 
+                                     float(RPlot[s])) >= rejectionRatio):
+            node.assignSplitPoint(-1)
+            # since splitpoint is not significant, 
+            # ignore this split and continue (reject both child nodes)
+            _cluster_tree(node, parentNode, localMaximaPoints, 
+                        RPlot, RPoints, min_cluster_size)
+            return
+ 
+    # remove clusters that are too small
+    if len(Node1.points) < min_cluster_size:
+        # cluster 1 is too small"
+        try:
+            Nodelist.remove((Node1, LocalMax1))
+        except Exception:
+            sys.exc_clear()
+    if len(Node2.points) < min_cluster_size:
+        # cluster 2 is too small
+        try:
+            Nodelist.remove((Node2, LocalMax2))
```

Comment:
Ahh yeah that makes things interesting.  My thoughts would be to change the conditionals to confirm the item exists before trying to remove it like so:

`if len(Node1.points) < min_cluster_size and Nodelist.count((Node1, LocalMax1)) > 0:`

Then maybe just drop the try/except that is in that statement - if there are exceptions we could let them bubble up?  

Also I'm not crazy about this method for exception handling because it seems like they are just eating them which can make things confusing if something is going wrong.  Maybe i'm just not seeing the big picture though - does @amyxzhang have any input?